File: src/Viasfora.LanguageServices/Viasfora.LanguageService.CSharp/ArgumentValidationTagger/SyntaxHelper.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Winterdom.Viasfora.LanguageService.CSharp.ArgumentValidationTagger {
  public static class SyntaxHelper {

    public static bool IsAnyIfArgumentThrowSyntaxStatement(SyntaxNode node) {
      if ( node == null )
        return false;

      var baseMethodDeclarationSyntax = node.FirstAncestorOrSelf<BaseMethodDeclarationSyntax>();
      if ( baseMethodDeclarationSyntax == null || baseMethodDeclarationSyntax == node)
        return false;

      if ( baseMethodDeclarationSyntax.Body == null || baseMethodDeclarationSyntax.Body.Statements == null )
        return false;

      foreach (var statementSyntax in baseMethodDeclarationSyntax.Body.Statements) {
        if (node.Span.IntersectsWith(statementSyntax.Span)) {
          return IsAnyIfArgumentThrowSyntaxStatement(statementSyntax);
        }
      }

      return false;
    }

    public static bool IsAnyIfArgumentThrowSyntaxStatement(StatementSyntax statementSyntax) {
      if ( statementSyntax == null ) 
        return false;

      var baseMethodDeclarationSyntax = statementSyntax.FirstAncestorOrSelf<BaseMethodDeclarationSyntax>();
      if ( baseMethodDeclarationSyntax == null )
        return false;

      bool isCtor = baseMethodDeclarationSyntax is ConstructorDeclarationSyntax;

      var parameterNames = baseMethodDeclarationSyntax.ParameterList.Parameters
        .Select(x => x.Identifier.ValueText);

      if ( !IsAnyIfArgumentThrowSyntaxStatementCore(statementSyntax, isCtor, parameterNames) )
        return false;

      if ( !CheckPreviouslyStatements(statementSyntax, isCtor, parameterNames) )
        return false;

      return true;
    }

    private static bool CheckPreviouslyStatements(StatementSyntax statementSyntax, bool isCtor, IEnumerable<string> parameterNames) {
      var blockSyntax = statementSyntax.Parent as BlockSyntax;
      if ( blockSyntax == null || blockSyntax.Statements.Count < 2 ) 
        return true;

      var statementList = blockSyntax.Statements;

      var startIdx = statementList.IndexOf(statementSyntax);
      for (var i=startIdx - 1; i>-1; --i ) {
        if ( !IsAnyIfArgumentThrowSyntaxStatementCore(statementList[i], isCtor, parameterNames) )
          return false;
      }

      return true;
    }

    private static bool IsAnyIfArgumentThrowSyntaxStatementCore(StatementSyntax statementSyntax, bool isCtor, IEnumerable<string> parameterNames) {
      var ifStatementSyntax = statementSyntax as IfStatementSyntax;
      if ( ifStatementSyntax != null )
        return IsIfArgumentThrowSyntaxStatementCore(ifStatementSyntax, parameterNames);

      if ( isCtor ) {
        if ( IsAssignmentCoalescThrowStatement(statementSyntax, parameterNames) )
          return true;
      }

      return false;
    }

    private static bool IsIfArgumentThrowSyntaxStatementCore(IfStatementSyntax ifStatementSyntax, IEnumerable<string> parameterNames) {
      // check if statement with only one statement in true part
      StatementSyntax innerIfStatementSyntax;
      if ( ifStatementSyntax.Statement is BlockSyntax blockSyntax ) {
        if ( blockSyntax.Statements.Count != 1 )
          return false;
        innerIfStatementSyntax = blockSyntax.Statements[0];
      } else
        innerIfStatementSyntax = ifStatementSyntax.Statement;

      // the inner if statement must be a throw Statement
      var throwStatementSyntax = innerIfStatementSyntax as ThrowStatementSyntax;
      if ( throwStatementSyntax == null )
        return false;

      // the if expression must contain at least one parameter(identifername)
      if ( !ExpressionContainsParameter(ifStatementSyntax.Condition, parameterNames) && !ExpressionContainsParameter(throwStatementSyntax.Expression, parameterNames) )
        return false;
      
      return true;
    }

    private static bool ExpressionContainsParameter(ExpressionSyntax expressionSyntax, IEnumerable<string> parameterNames) {
      if ( expressionSyntax == null ) 
        return false;

      // checks if the givven expression contains any ParameterNames. 
      // Skips MemberAccessExpressions like `this.XXX`

      var identifiers = expressionSyntax.DescendantNodesAndSelf()
        .OfType<IdentifierNameSyntax>()
      ;

      return identifiers.Any(x => {
        var memberAccessExpressionSyntax = x.FirstAncestorOrSelf<MemberAccessExpressionSyntax>();
        if ( memberAccessExpressionSyntax != null && memberAccessExpressionSyntax.SpanStart >= expressionSyntax.SpanStart ) {
          if (memberAccessExpressionSyntax.Expression != x)
            return false;
        }

        var invocationExpression = x.FirstAncestorOrSelf<InvocationExpressionSyntax>();
        if ( invocationExpression != null && invocationExpression.SpanStart >= expressionSyntax.SpanStart )
          return false;

        return parameterNames.Contains(x.Identifier.ValueText);
      });
    }

    private static bool IsAssignmentCoalescThrowStatement(StatementSyntax statementSyntax, IEnumerable<string> parameterNames) {
      var expressionStatementSyntax = statementSyntax as ExpressionStatementSyntax;
      if ( expressionStatementSyntax == null )
        return false;

      var assignmentExpressionSyntax = expressionStatementSyntax.Expression as AssignmentExpressionSyntax;
      if ( assignmentExpressionSyntax == null )
        return false;

      var binaryExpressionSyntax = assignmentExpressionSyntax.Right as BinaryExpressionSyntax;
      if ( binaryExpressionSyntax == null )
        return false;

      if ( binaryExpressionSyntax.Kind() != Microsoft.CodeAnalysis.CSharp.SyntaxKind.CoalesceExpression )
        return false;

      var leftIdentifierNameSyntax = binaryExpressionSyntax.Left as IdentifierNameSyntax;
      if ( leftIdentifierNameSyntax == null )
        return false;
      if ( !parameterNames.Contains(leftIdentifierNameSyntax.Identifier.ValueText) )
        return false;

      // skip check right throw ...
      // `this.xyz = xyz ?? xyzDefault;` should be also a valid expression 

      return true;
    }

  }
}

File: src/Viasfora.LanguageServices/Viasfora.LanguageService.Basic/ArgumentValidationTagger/SyntaxHelper.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace ArgumentValidationTagger

  Public Class SyntaxHelper


    Public Shared Function IsAnyIfArgumentThrowSyntaxStatement(node As SyntaxNode) As Boolean
      If node Is Nothing Then
        Return False
      End If

      Dim methodBlockBaseSyntax = node.FirstAncestorOrSelf(Of MethodBlockBaseSyntax)()
      If methodBlockBaseSyntax Is Nothing OrElse methodBlockBaseSyntax Is node Then
        Return False
      End If

      Dim accessorBlockSyntax = TryCast(methodBlockBaseSyntax, AccessorBlockSyntax)
      If accessorBlockSyntax IsNot Nothing Then
        ' Skip Properties
        Return False
      End If

      For Each statementSyntax In methodBlockBaseSyntax.Statements
        If node.Span.IntersectsWith(statementSyntax.Span) Then
          Return IsAnyIfArgumentThrowSyntaxStatement(statementSyntax)
        End If
      Next

      Return False
    End Function

    Public Shared Function IsAnyIfArgumentThrowSyntaxStatement(statementSyntax As StatementSyntax) As Boolean
      If statementSyntax Is Nothing Then
        Return False
      End If

      Dim methodBlockBaseSyntax = statementSyntax.FirstAncestorOrSelf(Of MethodBlockBaseSyntax)()
      If methodBlockBaseSyntax Is Nothing Then
        Return False
      End If

      Dim methodBaseSyntax = methodBlockBaseSyntax.BlockStatement
      If methodBaseSyntax Is Nothing Then
        Return False
      End If

      If methodBaseSyntax.ParameterList Is Nothing Then
        Return False
      End If

      Dim parameterNames = methodBaseSyntax.ParameterList.Parameters _
        .Select(Function(x) x.Identifier.Identifier.ValueText)

      If Not IsAnyIfArgumentThrowSyntaxStatementCore(statementSyntax, parameterNames) Then
        Return False
      End If

      If Not CheckPreviouslyStatements(statementSyntax, methodBlockBaseSyntax.Statements, parameterNames) Then
        Return False
      End If

      Return True
    End Function

    Friend Shared Function IsAnyIfArgumentThrowSyntaxStatementCore(statementSyntax As StatementSyntax, parameterNames As IEnumerable(Of String)) As Boolean
      Dim multiLineIfBlockSyntax = TryCast(statementSyntax, MultiLineIfBlockSyntax)
      If multiLineIfBlockSyntax IsNot Nothing Then
        Return IsIfArgumentThrowSyntaxStatementCore(multiLineIfBlockSyntax, parameterNames)
      End If

      Dim singleLineIfStatementSyntax = TryCast(statementSyntax, SingleLineIfStatementSyntax)
      If singleLineIfStatementSyntax IsNot Nothing Then
        Return IsIfArgumentThrowSyntaxStatementCore(singleLineIfStatementSyntax, parameterNames)
      End If

      Return False
    End Function

    Private Shared Function IsIfArgumentThrowSyntaxStatementCore(multiLineIfBlockSyntax As MultiLineIfBlockSyntax, parameterNames As IEnumerable(Of String)) As Boolean
      If multiLineIfBlockSyntax.Statements.Count <> 1 Then
        Return False
      End If

      Return IsIfArgumentThrowSyntaxStatementCore(multiLineIfBlockSyntax.IfStatement.Condition, multiLineIfBlockSyntax.Statements(0), parameterNames)
    End Function

    Private Shared Function IsIfArgumentThrowSyntaxStatementCore(singleLineIfStatementSyntax As SingleLineIfStatementSyntax, parameterNames As IEnumerable(Of String)) As Boolean
      If singleLineIfStatementSyntax.Statements.Count <> 1 Then
        Return False
      End If

      Return IsIfArgumentThrowSyntaxStatementCore(singleLineIfStatementSyntax.Condition, singleLineIfStatementSyntax.Statements(0), parameterNames)
    End Function

    Private Shared Function IsIfArgumentThrowSyntaxStatementCore(condition As ExpressionSyntax, trueStatementSyntax As StatementSyntax, parameterNames As IEnumerable(Of String)) As Boolean
      If condition Is Nothing Then
        Return False
      End If

      If trueStatementSyntax Is Nothing Then
        Return False
      End If

      Dim throwStatement = TryCast(trueStatementSyntax, ThrowStatementSyntax)
      If throwStatement Is Nothing Then
        Return False
      End If

      ' the if expression must contain at least one parameter(identifername)
      If Not ExpressionContainsParameter(condition, parameterNames) AndAlso
          Not ExpressionContainsParameter(throwStatement.Expression, parameterNames) Then

        Return False
      End If

      Return True
    End Function

    Private Shared Function ExpressionContainsParameter(expressionSyntax As ExpressionSyntax, parameterNames As IEnumerable(Of String)) As Boolean
      If expressionSyntax Is Nothing Then
        Return False
      End If

      Dim identifiers = expressionSyntax.DescendantNodesAndSelf() _
        .OfType(Of IdentifierNameSyntax)

      Return identifiers.Any(
        Function(x)
          Dim memberAccessExpressionSyntax = x.FirstAncestorOrSelf(Of MemberAccessExpressionSyntax)()
          If memberAccessExpressionSyntax IsNot Nothing AndAlso memberAccessExpressionSyntax.SpanStart >= expressionSyntax.SpanStart Then
            If memberAccessExpressionSyntax.Expression IsNot x Then
              Return False
            End If
          End If

          Dim invocationExpression = x.FirstAncestorOrSelf(Of InvocationExpressionSyntax)()
          If invocationExpression IsNot Nothing AndAlso invocationExpression.SpanStart >= expressionSyntax.SpanStart Then
            Return False
          End If

          Return parameterNames.Contains(x.Identifier.ValueText, StringComparer.OrdinalIgnoreCase)
        End Function
      )
    End Function

    Private Shared Function CheckPreviouslyStatements(statementSyntax As StatementSyntax, parentStatementList As SyntaxList(Of StatementSyntax), parameterNames As IEnumerable(Of String)) As Boolean
      If parentStatementList.Count < 2 Then
        Return True
      End If

      Dim startIdx = parentStatementList.IndexOf(statementSyntax)
      For i = startIdx - 1 To 0 Step -1
        If Not IsAnyIfArgumentThrowSyntaxStatementCore(parentStatementList(i), parameterNames) Then
          Return False
        End If
      Next

      Return True
    End Function

  End Class

End Namespace

